File: steps/python/adoption_check.step.py
# steps/python/adoption_check.step.py
import sys
import os
sys.path.append(os.path.join(os.path.dirname(__file__), '..'))

from services import pet_store
import asyncio

config = {
    "type": "event",
    "name": "PyBackgroundCheck",
    "description": "Validates adopter history and pet availability in parallel with application summarizer",
    "subscribes": ["py.adoption.applied"],
    "emits": ["py.adoption.background.complete"],
    "flows": ["python-adoptions"]
}

async def handler(event, ctx=None):
    if not ctx:
        return
        
    application_id = (event or {}).get("applicationId")
    pet_id = (event or {}).get("petId")
    adopter_name = (event or {}).get("adopterName")
    adopter_email = (event or {}).get("adopterEmail")
    
    ctx.logger.info("🔍 Running background check", {
        "applicationId": application_id,
        "petId": pet_id,
        "adopterName": adopter_name
    })

    # Update stream to show checking phase
    # Background check starting - no stream update needed

    # Simulate background check logic
    check_result = "passed"
    check_details = "All checks passed successfully"

    try:
        # Check pet availability
        pet = pet_store.get(pet_id)
        if not pet:
            check_result = "failed"
            check_details = "Pet not found"
        elif pet.get("status") != "available":
            check_result = "failed"
            check_details = "Pet is not available for adoption"

        # Check adopter email (simulate spam detection)
        if adopter_email and "spam" in adopter_email.lower():
            check_result = "failed"
            check_details = "Email flagged as potential spam"

        # Check adopter name (simulate basic validation)
        if adopter_name and len(adopter_name) < 3:
            check_result = "failed"
            check_details = "Adopter name too short"

        # Simulate processing delay
        await asyncio.sleep(0.5)

        ctx.logger.info("Background check completed", {
            "applicationId": application_id,
            "petId": pet_id,
            "checkResult": check_result,
            "checkDetails": check_details
        })

        # Emit completion event with check results
        await ctx.emit({
            "topic": "py.adoption.background.complete",
            "data": {
                "applicationId": application_id,
                "petId": pet_id,
                "adopterName": adopter_name,
                "adopterEmail": adopter_email,
                "checkResult": check_result,
                "checkDetails": check_details,
            }
        })

    except Exception as error:
        ctx.logger.error("Background check failed", {
            "applicationId": application_id,
            "petId": pet_id,
            "error": str(error)
        })
        
        await ctx.emit({
            "topic": "py.adoption.background.complete",
            "data": {
                "applicationId": application_id,
                "petId": pet_id,
                "adopterName": adopter_name,
                "adopterEmail": adopter_email,
                "checkResult": "error",
                "checkDetails": f"Check failed: {str(error)}",
            }
        })

File: steps/javascript/adoption.check.step.js
// steps/javascript/adoption.check.step.js
const { get } = require('./js-store');

exports.config = {
  type: 'event',
  name: 'JsBackgroundCheck',
  description: 'Validates adopter history and pet availability in parallel with application summarizer',
  subscribes: ['js.adoption.applied'],
  emits: ['js.adoption.background.complete'],
  flows: ['javascript-adoptions'],
};

exports.handler = async (event, { emit, logger, streams, traceId }) => {
  const { applicationId, petId, adopterName, adopterEmail } = event || {};

  logger.info('🔍 Running background check', { applicationId, petId, adopterName });

  // Update stream to show checking phase
  if (streams?.adoptions && traceId) {
    await streams.adoptions.set(traceId, 'status', {
      entityId: applicationId,
      type: 'application',
      phase: 'checking',
      message: 'Running background check and generating summary',
      timestamp: Date.now(),
      data: { petId, adopterName }
    });
  }

  // Simulate background check logic
  let checkResult = 'passed';
  let checkDetails = 'All checks passed successfully';

  try {
    // Check pet availability
    const pet = get(petId);
    if (!pet) {
      checkResult = 'failed';
      checkDetails = 'Pet not found';
    } else if (pet.status !== 'available') {
      checkResult = 'failed';
      checkDetails = 'Pet is not available for adoption';
    }

    // Check adopter email (simulate spam detection)
    if (adopterEmail && adopterEmail.toLowerCase().includes('spam')) {
      checkResult = 'failed';
      checkDetails = 'Email flagged as potential spam';
    }

    // Check adopter name (simulate basic validation)
    if (adopterName && adopterName.length < 3) {
      checkResult = 'failed';
      checkDetails = 'Adopter name too short';
    }

    // Simulate processing delay
    await new Promise(resolve => setTimeout(resolve, 500));

    logger.info('Background check completed', { 
      applicationId, 
      petId, 
      checkResult, 
      checkDetails 
    });

    // Emit completion event with check results
    await emit({
      topic: 'js.adoption.background.complete',
      data: {
        applicationId,
        petId,
        adopterName,
        adopterEmail,
        checkResult,
        checkDetails,
        traceId
      }
    });

  } catch (error) {
    logger.error('Background check failed', { applicationId, petId, error: error.message });
    
    await emit({
      topic: 'js.adoption.background.complete',
      data: {
        applicationId,
        petId,
        adopterName,
        adopterEmail,
        checkResult: 'error',
        checkDetails: `Check failed: ${error.message}`,
        traceId
      }
    });
  }
};
